### PR TITLE
Improvements in High-Level API and Poe Tasks

### LINKED ISSUE #1054: Allow number of episodes per test step to be configured in high-level API
At present, this is hard-coded to the number of test environments, i.e. one episode per test environment, because this is how it was implemented in all examples, but obviously, it should be configurable.

I propose to add a configuration item with the total number of episodes to collect per episode. If it is not a multiple of the number of test environments, then we round up to the nearest multiple.

### PR DESCRIPTION
**Pull request by opcode81**
- [X] I have added the correct label(s) to this Pull Request or linked the relevant issue(s)
- [X] I have provided a description of the changes in this Pull Request
- [X] I have added documentation for my changes
- [ ] If applicable, I have added tests to cover my changes.
- [X] I have reformatted the code using `poe format` 
- [X] I have checked style and types with `poe lint` and `poe type-check`
- [ ] (Optional) I ran tests locally with `poe test` 
(or a subset of them with `poe test-reduced`) ,and they pass
- [ ] (Optional) I have tested that documentation builds correctly with `poe doc-build`

* Add an option to SamplingConfig which allows to configure number of test episodes
* Make OptimizerFactory more flexible, adding method `create_optimizer_for_params`
* Fix AutoAlphaFactoryDefault using hard-coded Adam optimizer
* Fix mypy issues that were platform/installation-dependent
* Limit scope of nbqa, resolving issues with files generated by old versions of the build

Fixes #1054

### CODE REVIEW

File: tianshou/highlevel/optim.py [hunk 1]
```diff
@@ -1,20 +1,30 @@
...
+    def create_optimizer(
+        self,
+        module: torch.nn.Module,
+        lr: float,
+    ) -> torch.optim.Optimizer:
+        return self.create_optimizer_for_params(module.parameters(), lr)
+
     @abstractmethod
-    def create_optimizer(self, module: torch.nn.Module, lr: float) -> torch.optim.Optimizer:
+    def create_optimizer_for_params(self, params: TParams, lr: float) -> torch.optim.Optimizer:
```
Comment (MischaPanch) [inline tianshou/highlevel/optim.py:27]: Would it have a downside to have a non-abstract `create_optimizer` here with the previous signature that uses `create_optimizer_for_params`? Then you wouldn't need the downstream changes while still keeping flexibility.

Using the `module.params` is a good default that works for most cases

File: tianshou/highlevel/params/alpha.py [hunk 2]
```diff
@@ -32,5 +32,5 @@ def create_auto_alpha(
     ) -> tuple[float, torch.Tensor, torch.optim.Optimizer]:
         target_entropy = float(-np.prod(envs.get_action_shape()))
         log_alpha = torch.zeros(1, requires_grad=True, device=device)
-        alpha_optim = torch.optim.Adam([log_alpha], lr=self.lr)
+        alpha_optim = optim_factory.create_optimizer_for_params([log_alpha], self.lr)
```
Comment (MischaPanch) [inline tianshou/highlevel/params/alpha.py:35]: I prefer using kwargs (i.e. `params=[log_alpha], lr=self.lr`), also safer against refactorings. Maybe force them in the new method?

You could also force all kwargs after `params`, that one kinda makes sense as positional arg

**Comment by opcode81** [comment]
Recommend to merge without squashing.

File: tianshou/highlevel/params/alpha.py (hunk 2 above)
Comment (opcode81) [reply tianshou/highlevel/params/alpha.py:35]: It is not reasonable to require kwargs for a two-arg method with two entirely different types.

File: tianshou/highlevel/optim.py (hunk 1 above)
Comment (opcode81) [reply tianshou/highlevel/optim.py:27]: I'm confused: `create_optimizer` still exists, is non-abstract, is implemented via `create_optimizer_for_params` and there are no downstream changes (in user code; only this module).

Comment (MischaPanch) [reply tianshou/highlevel/optim.py:27]: Sorry, I glanced over the diff too fast and misread it. All good, merging

**Comment by MischaPanch** [comment]
The windows CI job fails b/c you changed pyproject and something about caching fails. I'll delete the current github caches before merging to master, that should fix it

**Review by MischaPanch — APPROVED** [review]
